src/index.js

```diff
@@ -520,6 +553,206 @@ class EventsCoreAPI {
+      throw error
+    }
+
+    if (publicKey) {
+      return publicKey
+    }
+
+    logger.info('public key for url %s not present in state lib cache, fetching from api..', pubKeyUrl)
+    // fetch from api and set in cache default expiry 24h
+    publicKey = this.__fetchPublicKey(pubKeyUrl)
+    await state.put(publicKeyFileName, { anObject: publicKey })
+    logger.info('fetched key set in the aio lib state')
+    return publicKey
+  }
+
+  /**
+   * Fetches public key using the cloud front public key url
+   * 
+   * @param {*} publicKeyUrl - cloud front public key url of format 
+   * https://d2wbnl47xxxxx.cloudfront.net/pub-key-1.pem
+   * @returns public key
+   */
+  __fetchPublicKey (publicKeyUrl) {
+    let pubKey
+    fetch(publicKeyUrl)
+      .then(res => res.text())
+      .then(text => {
+        logger.debug('successfully fetched the public key %s from api', text)
+        pubKey = text
+      })
+      .catch(err => {
+        logger.error('error fetching the public key from api due to => %s', err.message)
+        throw err
+      })
+    return pubKey
+  }
+
+  /**
+   * Digital Signature Verification Helper
+   * 
+   * @param {*} digitalSignature1 - I/O Events generated digital signature1
+   * @param {*} digitalSignature2 - I/O Events generated digital signature2
+   * @param {*} signedPayload - I/O Events proper signed payload
+   * @param {*} pubKey1Pem - I/O Events generated PEM encoded public key1
+   * @param {*} pubKey2Pem - I/O Events generated PEM encoded public key1
+   * @param {*} recipientClientId - target recipient client id
+   * @returns {boolean} true if either signatures are valid or false
+   */
+  __verifySignature (digitalSignature1, digitalSignature2, signedPayload, pubKey1Pem, pubKey2Pem, recipientClientId) {
+    let result1, result2, publicKey1, publicKey2
+    try {
+      publicKey1 = this.__createCryptoPublicKey(pubKey1Pem)
+      result1 = this.__cryptoVerify(digitalSignature1, publicKey1, signedPayload)
+      if (result1) {
+        return result1
+      } 
+
+      publicKey2 = this.__createCryptoPublicKey(pubKey2Pem)
+      result2 = this.__cryptoVerify(digitalSignature2, publicKey2, signedPayload)
+      return result2
+    } catch (error) {
+      logger.error('error occured while verifying digital signature for client id %s due to %s ', recipientClientId, error.message)
+      return this.__genErrorResponse(500, 'error occured while verifying digital signature')
+    }
+  }
+
+  /**
+   * Crypto verifies the signature of the I/O Events signed payload
+   * 
+   * @param {*} signature I/O Events digital signature
+   * @param {*} pubKey I/O Events public key
+   * @param {*} signedPayload I/O Events signed payload 
+   * @returns {boolean} true or false
+   */
+  __cryptoVerify (signature, pubKey, signedPayload) {
+    return crypto.verify('rsa-sha256', new TextEncoder().encode(signedPayload), pubKey, Buffer.from(signature, 'base64'))
+  }
+
+  __createCryptoPublicKey (publicKey) {
```

Comment:
Shall we refactor all signature related helpers in a separate file? This `index.js` is getting too bulky and will be harder to maintain in the future.

Reply:
yes. I felt the same and was going to work on it after fixing the tests.